File: engine/web-ui/src/lib/db-modules/codegen-analytics.ts
import { randomUUID } from "crypto";
import { and, desc, eq } from "drizzle-orm";
import drizzleDb from "./drizzle";
import {
    codegenGeneratedFiles,
    codegenInputEntities,
    codegenRunMetrics,
    codegenRuns,
} from "./schema";

export type CodegenRunStatus = "queued" | "running" | "completed" | "failed" | "cancelled";
export type CodegenRunSourceType = "manual" | "derived_causal" | "follow_up" | "imported";
export type CodegenMetricType = "text" | "number" | "boolean" | "json";

export type CreateCodegenRunInput = {
    id?: string;
    projectId?: string;
    componentId?: string;
    causalProjectDocumentId?: string;
    sourceType?: CodegenRunSourceType;
    status?: CodegenRunStatus;
    model?: string;
    inputPrompt?: string;
    startedAt?: string;
};

export type RecordCodegenInputEntity = {
    id?: string;
    entityName: string;
    sourceCausalId?: string;
    sourceHead?: string;
    sourceRelationship?: string;
    sourceTail?: string;
    sourceDetail?: string;
};

export type RecordCodegenGeneratedFile = {
    id?: string;
    entityName: string;
    filePath: string;
    language?: string;
    fileSizeBytes?: number;
    generationOrder?: number;
};

export type UpsertCodegenMetricInput = {
    key: string;
    value: string | number | boolean | Record<string, unknown> | Array<unknown>;
    type?: CodegenMetricType;
};

export type CodegenRunRecord = {
    id: string;
    projectId: string | null;
    componentId: string | null;
    causalProjectDocumentId: string | null;
    sourceType: CodegenRunSourceType;
    status: CodegenRunStatus;
    model: string | null;
    inputPrompt: string | null;
    startedAt: string | null;
    finishedAt: string | null;
    durationMs: number | null;
    inputEntityCount: number;
    generatedEntityCount: number;
    generatedFileCount: number;
    errorMessage: string | null;
    createdAt: string;
    updatedAt: string;
};

function nowIso(): string {
    return new Date().toISOString();
}

function normalizeRunStatus(status: string | undefined): CodegenRunStatus {
    if (status === "running" || status === "completed" || status === "failed" || status === "cancelled") {
        return status;
    }
    return "queued";
}

function normalizeSourceType(sourceType: string | undefined): CodegenRunSourceType {
    if (sourceType === "derived_causal" || sourceType === "follow_up" || sourceType === "imported") {
        return sourceType;
    }
    return "manual";
}

function normalizeMetricType(metricType: string | undefined): CodegenMetricType {
    if (metricType === "number" || metricType === "boolean" || metricType === "json") {
        return metricType;
    }
    return "text";
}

function metricValueToText(value: UpsertCodegenMetricInput["value"], type: CodegenMetricType): string {
    if (type === "json") {
        return JSON.stringify(value);
    }

    if (type === "boolean") {
        return String(Boolean(value));
    }

    if (type === "number") {
        const numeric = typeof value === "number" ? value : Number(value);
        return Number.isFinite(numeric) ? String(numeric) : "0";
    }

    if (typeof value === "string") {
        return value;
    }

    return JSON.stringify(value);
}

function toCodegenRunRecord(row: {
    id: string;
    projectId: string | null;
    componentId: string | null;
    causalProjectDocumentId: string | null;
    sourceType: string;
    status: string;
    model: string | null;
    inputPrompt: string | null;
    startedAt: string | null;
    finishedAt: string | null;
    durationMs: number | null;
    inputEntityCount: number;
    generatedEntityCount: number;
    generatedFileCount: number;
    errorMessage: string | null;
    createdAt: string;
    updatedAt: string;
}): CodegenRunRecord {
    return {
        id: row.id,
        projectId: row.projectId,
        componentId: row.componentId,
        causalProjectDocumentId: row.causalProjectDocumentId,
        sourceType: normalizeSourceType(row.sourceType),
        status: normalizeRunStatus(row.status),
        model: row.model,
        inputPrompt: row.inputPrompt,
        startedAt: row.startedAt,
        finishedAt: row.finishedAt,
        durationMs: row.durationMs,
        inputEntityCount: row.inputEntityCount,
        generatedEntityCount: row.generatedEntityCount,
        generatedFileCount: row.generatedFileCount,
        errorMessage: row.errorMessage,
        createdAt: row.createdAt,
        updatedAt: row.updatedAt,
    };
}

export function createCodegenRun(input: CreateCodegenRunInput): CodegenRunRecord {
    const createdAt = nowIso();
    const runId = (input.id ?? randomUUID()).trim();
    if (!runId) {
        throw new Error("Codegen run id is required.");
    }

    const sourceType = normalizeSourceType(input.sourceType);
    const status = normalizeRunStatus(input.status);

    drizzleDb
        .insert(codegenRuns)
        .values({
            id: runId,
            projectId: input.projectId?.trim() || null,
            componentId: input.componentId?.trim() || null,
            causalProjectDocumentId: input.causalProjectDocumentId?.trim() || null,
            sourceType,
            status,
            model: input.model?.trim() || null,
            inputPrompt: input.inputPrompt ?? null,
            startedAt: input.startedAt || (status === "running" ? createdAt : null),
            finishedAt: null,
            durationMs: null,
            inputEntityCount: 0,
            generatedEntityCount: 0,
            generatedFileCount: 0,
            errorMessage: null,
            createdAt,
            updatedAt: createdAt,
        })
        .run();

    return getCodegenRunById(runId) as CodegenRunRecord;
}

export function getCodegenRunById(runId: string): CodegenRunRecord | null {
    const trimmed = runId.trim();
    if (!trimmed) {
        return null;
    }

    const row = drizzleDb
        .select({
            id: codegenRuns.id,
            projectId: codegenRuns.projectId,
            componentId: codegenRuns.componentId,
            causalProjectDocumentId: codegenRuns.causalProjectDocumentId,
            sourceType: codegenRuns.sourceType,
            status: codegenRuns.status,
            model: codegenRuns.model,
            inputPrompt: codegenRuns.inputPrompt,
            startedAt: codegenRuns.startedAt,
            finishedAt: codegenRuns.finishedAt,
            durationMs: codegenRuns.durationMs,
            inputEntityCount: codegenRuns.inputEntityCount,
            generatedEntityCount: codegenRuns.generatedEntityCount,
            generatedFileCount: codegenRuns.generatedFileCount,
            errorMessage: codegenRuns.errorMessage,
            createdAt: codegenRuns.createdAt,
            updatedAt: codegenRuns.updatedAt,
        })
        .from(codegenRuns)
        .where(eq(codegenRuns.id, trimmed))
        .get();

    return row ? toCodegenRunRecord(row) : null;
}

export function listCodegenRunsForComponent(componentId: string, limit = 50): CodegenRunRecord[] {
    const trimmed = componentId.trim();
    if (!trimmed) {
        return [];
    }

    const safeLimit = Math.max(1, Math.min(limit, 1000));
    const rows = drizzleDb
        .select({
            id: codegenRuns.id,
            projectId: codegenRuns.projectId,
            componentId: codegenRuns.componentId,
            causalProjectDocumentId: codegenRuns.causalProjectDocumentId,
            sourceType: codegenRuns.sourceType,
            status: codegenRuns.status,
            model: codegenRuns.model,
            inputPrompt: codegenRuns.inputPrompt,
            startedAt: codegenRuns.startedAt,
            finishedAt: codegenRuns.finishedAt,
            durationMs: codegenRuns.durationMs,
            inputEntityCount: codegenRuns.inputEntityCount,
            generatedEntityCount: codegenRuns.generatedEntityCount,
            generatedFileCount: codegenRuns.generatedFileCount,
            errorMessage: codegenRuns.errorMessage,
            createdAt: codegenRuns.createdAt,
            updatedAt: codegenRuns.updatedAt,
        })
        .from(codegenRuns)
        .where(eq(codegenRuns.componentId, trimmed))
        .orderBy(desc(codegenRuns.createdAt))
        .limit(safeLimit)
        .all();

    return rows.map(toCodegenRunRecord);
}

export function markCodegenRunRunning(runId: string, startedAt?: string): boolean {
    const trimmed = runId.trim();
    if (!trimmed) {
        return false;
    }

    const now = nowIso();
    const result = drizzleDb
        .update(codegenRuns)
        .set({
            status: "running",
            startedAt: startedAt || now,
            updatedAt: now,
            errorMessage: null,
        })
        .where(eq(codegenRuns.id, trimmed))
        .run();

    return (result.changes ?? 0) > 0;
}

export function markCodegenRunCompleted(
    runId: string,
    input?: {
        finishedAt?: string;
        durationMs?: number;
        generatedEntityCount?: number;
        generatedFileCount?: number;
    },
): boolean {
    const trimmed = runId.trim();
    if (!trimmed) {
        return false;
    }

    const now = nowIso();
    const current = getCodegenRunById(trimmed);
    if (!current) {
        return false;
    }

    const finishedAt = input?.finishedAt || now;
    let durationMs = input?.durationMs;
    if (durationMs == null && current.startedAt) {
        const startedMillis = Date.parse(current.startedAt);
        const finishedMillis = Date.parse(finishedAt);
        if (Number.isFinite(startedMillis) && Number.isFinite(finishedMillis)) {
            durationMs = Math.max(0, finishedMillis - startedMillis);
        }
    }

    const generatedFileCount =
        input?.generatedFileCount ??
        drizzleDb
            .select({ count: codegenGeneratedFiles.id })
            .from(codegenGeneratedFiles)
            .where(eq(codegenGeneratedFiles.runId, trimmed))
            .all().length;

    drizzleDb
        .update(codegenRuns)
        .set({
            status: "completed",
            finishedAt,
            durationMs: durationMs ?? null,
            generatedEntityCount: input?.generatedEntityCount ?? current.generatedEntityCount,
            generatedFileCount,
            updatedAt: now,
            errorMessage: null,
        })
        .where(eq(codegenRuns.id, trimmed))
        .run();

    return true;
}

export function markCodegenRunFailed(runId: string, errorMessage: string, finishedAt?: string): boolean {
    const trimmed = runId.trim();
    if (!trimmed) {
        return false;
    }

    const now = nowIso();
    const result = drizzleDb
        .update(codegenRuns)
        .set({
            status: "failed",
            finishedAt: finishedAt || now,
            errorMessage: errorMessage.trim() || "Unknown codegen failure.",
            updatedAt: now,
        })
        .where(eq(codegenRuns.id, trimmed))
        .run();

    return (result.changes ?? 0) > 0;
}

export function saveCodegenInputEntities(runId: string, entities: RecordCodegenInputEntity[], replace = true): number {
    const trimmed = runId.trim();
    if (!trimmed) {
        throw new Error("runId is required.");
    }

    const now = nowIso();

    return drizzleDb.transaction((tx) => {
        if (replace) {
            tx.delete(codegenInputEntities).where(eq(codegenInputEntities.runId, trimmed)).run();
        }

        let inserted = 0;
        for (const entity of entities) {
            const name = entity.entityName.trim();
            if (!name) {
                continue;
            }

            tx
                .insert(codegenInputEntities)
                .values({
                    id: entity.id?.trim() || randomUUID(),
                    runId: trimmed,
                    entityName: name,
                    sourceCausalId: entity.sourceCausalId?.trim() || null,
                    sourceHead: entity.sourceHead?.trim() || null,
                    sourceRelationship: entity.sourceRelationship?.trim() || null,
                    sourceTail: entity.sourceTail?.trim() || null,
                    sourceDetail: entity.sourceDetail?.trim() || null,
                    createdAt: now,
                })
                .onConflictDoUpdate({
                    target: [codegenInputEntities.runId, codegenInputEntities.entityName],
                    set: {
                        sourceCausalId: entity.sourceCausalId?.trim() || null,
                        sourceHead: entity.sourceHead?.trim() || null,
                        sourceRelationship: entity.sourceRelationship?.trim() || null,
                        sourceTail: entity.sourceTail?.trim() || null,
                        sourceDetail: entity.sourceDetail?.trim() || null,
                        createdAt: now,
                    },
                })
                .run();

            inserted += 1;
        }

        tx
            .update(codegenRuns)
            .set({
                inputEntityCount: inserted,
                updatedAt: now,
            })
            .where(eq(codegenRuns.id, trimmed))
            .run();

        return inserted;
    });
}

export function saveCodegenGeneratedFiles(runId: string, files: RecordCodegenGeneratedFile[], replace = true): number {
    const trimmed = runId.trim();
    if (!trimmed) {
        throw new Error("runId is required.");
    }

    const now = nowIso();

    return drizzleDb.transaction((tx) => {
        if (replace) {
            tx.delete(codegenGeneratedFiles).where(eq(codegenGeneratedFiles.runId, trimmed)).run();
        }

        let inserted = 0;
        const uniqueEntities = new Set<string>();

        for (let index = 0; index < files.length; index += 1) {
            const file = files[index];
            const entityName = file.entityName.trim();
            const filePath = file.filePath.trim();
            if (!entityName || !filePath) {
                continue;
            }

            tx
                .insert(codegenGeneratedFiles)
                .values({
                    id: file.id?.trim() || randomUUID(),
                    runId: trimmed,
                    entityName,
                    filePath,
                    language: file.language?.trim() || null,
                    fileSizeBytes: typeof file.fileSizeBytes === "number" ? file.fileSizeBytes : null,
                    generationOrder: typeof file.generationOrder === "number" ? file.generationOrder : index,
                    createdAt: now,
                })
                .onConflictDoUpdate({
                    target: [codegenGeneratedFiles.runId, codegenGeneratedFiles.filePath],
                    set: {
                        entityName,
                        language: file.language?.trim() || null,
                        fileSizeBytes: typeof file.fileSizeBytes === "number" ? file.fileSizeBytes : null,
                        generationOrder: typeof file.generationOrder === "number" ? file.generationOrder : index,
                        createdAt: now,
                    },
                })
                .run();

            uniqueEntities.add(entityName);
            inserted += 1;
        }

        tx
            .update(codegenRuns)
            .set({
                generatedFileCount: inserted,
                generatedEntityCount: uniqueEntities.size,
                updatedAt: now,
            })
            .where(eq(codegenRuns.id, trimmed))
            .run();

        return inserted;
    });
}

export function upsertCodegenMetrics(runId: string, metrics: UpsertCodegenMetricInput[]): number {
    const trimmed = runId.trim();
    if (!trimmed) {
        throw new Error("runId is required.");
    }

    const now = nowIso();
    let upserted = 0;

    drizzleDb.transaction((tx) => {
        for (const metric of metrics) {
            const key = metric.key.trim();
            if (!key) {
                continue;
            }

            const metricType = normalizeMetricType(metric.type);
            const metricValue = metricValueToText(metric.value, metricType);

            tx
                .insert(codegenRunMetrics)
                .values({
                    id: `${trimmed}:${key}`,
                    runId: trimmed,
                    metricKey: key,
                    metricType,
                    metricValue,
                    createdAt: now,
                })
                .onConflictDoUpdate({
                    target: [codegenRunMetrics.runId, codegenRunMetrics.metricKey],
                    set: {
                        metricType,
                        metricValue,
                        createdAt: now,
                    },
                })
                .run();

            upserted += 1;
        }

        tx
            .update(codegenRuns)
            .set({
                updatedAt: now,
            })
            .where(eq(codegenRuns.id, trimmed))
            .run();
    });

    return upserted;
}

export function listCodegenGeneratedFiles(runId: string): Array<{
    id: string;
    runId: string;
    entityName: string;
    filePath: string;
    language: string | null;
    fileSizeBytes: number | null;
    generationOrder: number;
    createdAt: string;
}> {
    const trimmed = runId.trim();
    if (!trimmed) {
        return [];
    }

    return drizzleDb
        .select({
            id: codegenGeneratedFiles.id,
            runId: codegenGeneratedFiles.runId,
            entityName: codegenGeneratedFiles.entityName,
            filePath: codegenGeneratedFiles.filePath,
            language: codegenGeneratedFiles.language,
            fileSizeBytes: codegenGeneratedFiles.fileSizeBytes,
            generationOrder: codegenGeneratedFiles.generationOrder,
            createdAt: codegenGeneratedFiles.createdAt,
        })
        .from(codegenGeneratedFiles)
        .where(eq(codegenGeneratedFiles.runId, trimmed))
        .orderBy(codegenGeneratedFiles.generationOrder)
        .all();
}

export function listCodegenMetrics(runId: string): Array<{
    key: string;
    type: CodegenMetricType;
    value: string;
}> {
    const trimmed = runId.trim();
    if (!trimmed) {
        return [];
    }

    const rows = drizzleDb
        .select({
            key: codegenRunMetrics.metricKey,
            type: codegenRunMetrics.metricType,
            value: codegenRunMetrics.metricValue,
        })
        .from(codegenRunMetrics)
        .where(eq(codegenRunMetrics.runId, trimmed))
        .orderBy(codegenRunMetrics.metricKey)
        .all();

    return rows.map((row) => ({
        key: row.key,
        type: normalizeMetricType(row.type),
        value: row.value,
    }));
}

export function listCodegenRuns(
    input?: {
        projectId?: string;
        componentId?: string;
        status?: CodegenRunStatus;
        limit?: number;
    },
): CodegenRunRecord[] {
    const safeLimit = Math.max(1, Math.min(input?.limit ?? 200, 1000));

    const predicates = [];
    if (input?.projectId?.trim()) {
        predicates.push(eq(codegenRuns.projectId, input.projectId.trim()));
    }
    if (input?.componentId?.trim()) {
        predicates.push(eq(codegenRuns.componentId, input.componentId.trim()));
    }
    if (input?.status) {
        predicates.push(eq(codegenRuns.status, input.status));
    }

    const rows = drizzleDb
        .select({
            id: codegenRuns.id,
            projectId: codegenRuns.projectId,
            componentId: codegenRuns.componentId,
            causalProjectDocumentId: codegenRuns.causalProjectDocumentId,
            sourceType: codegenRuns.sourceType,
            status: codegenRuns.status,
            model: codegenRuns.model,
            inputPrompt: codegenRuns.inputPrompt,
            startedAt: codegenRuns.startedAt,
            finishedAt: codegenRuns.finishedAt,
            durationMs: codegenRuns.durationMs,
            inputEntityCount: codegenRuns.inputEntityCount,
            generatedEntityCount: codegenRuns.generatedEntityCount,
            generatedFileCount: codegenRuns.generatedFileCount,
            errorMessage: codegenRuns.errorMessage,
            createdAt: codegenRuns.createdAt,
            updatedAt: codegenRuns.updatedAt,
        })
        .from(codegenRuns)
        .where(predicates.length > 0 ? and(...predicates) : undefined)
        .orderBy(desc(codegenRuns.createdAt))
        .limit(safeLimit)
        .all();

    return rows.map(toCodegenRunRecord);
}

File: engine/backend/app/services/codegen_analytics.py
from __future__ import annotations

import os
import json
import sqlite3
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from ...infra.paths import ROOT_DIR


CODEGEN_ANALYTICS_TABLES = (
    "fact_codegen_runs",
    "fact_codegen_generated_files",
    "fact_codegen_input_entities",
    "fact_codegen_metrics",
)

CODEGEN_SOURCE_TABLES = (
    "codegen_runs",
    "codegen_generated_files",
    "codegen_input_entities",
    "codegen_run_metrics",
)


class CodegenAnalyticsError(RuntimeError):
    """Raised when codegen analytics refresh or queries fail."""


def _resolve_source_db_path(db_path: str | None = None) -> Path:
    configured = (db_path or os.getenv("CODEGEN_ANALYTICS_SOURCE_DB_PATH") or "").strip()
    if configured:
        return Path(configured).expanduser().resolve()

    candidates = [
        ROOT_DIR / "Engine" / "web-ui" / "local.db",
        ROOT_DIR / "engine" / "web-ui" / "local.db",
    ]

    for candidate in candidates:
        if candidate.exists():
            return candidate.resolve()

    return candidates[0].resolve()


def _resolve_analytics_db_path(db_path: str | None = None) -> Path:
    configured = (db_path or os.getenv("CODEGEN_ANALYTICS_DB_PATH") or "").strip()
    if configured:
        return Path(configured).expanduser().resolve()

    return (ROOT_DIR / "Engine" / "output" / "analytics" / "codegen_analytics.db").resolve()


def _table_exists(connection: sqlite3.Connection, table_name: str, schema_name: str | None = None) -> bool:
    source = f"{schema_name}.sqlite_master" if schema_name else "sqlite_master"
    row = connection.execute(
        f"SELECT 1 FROM {source} WHERE type='table' AND name = ? LIMIT 1",
        (table_name,),
    ).fetchone()
    return row is not None


def _create_codegen_analytics_schema(connection: sqlite3.Connection) -> None:
    connection.executescript(
        """
        CREATE TABLE IF NOT EXISTS fact_codegen_runs (
          run_id TEXT PRIMARY KEY NOT NULL,
          project_id TEXT,
          project_name TEXT,
          component_id TEXT,
          component_title TEXT,
          component_category TEXT,
          causal_project_document_id TEXT,
          source_type TEXT NOT NULL,
          status TEXT NOT NULL,
          model TEXT,
          started_at TEXT,
          finished_at TEXT,
          duration_ms INTEGER,
          duration_seconds REAL,
          input_entity_count INTEGER NOT NULL,
          generated_entity_count INTEGER NOT NULL,
          generated_file_count INTEGER NOT NULL,
          error_message TEXT,
          created_at TEXT NOT NULL,
          updated_at TEXT NOT NULL,
          is_success INTEGER NOT NULL DEFAULT 0
        );
        CREATE INDEX IF NOT EXISTS idx_fact_codegen_runs_project_id ON fact_codegen_runs(project_id);
        CREATE INDEX IF NOT EXISTS idx_fact_codegen_runs_component_id ON fact_codegen_runs(component_id);
        CREATE INDEX IF NOT EXISTS idx_fact_codegen_runs_status ON fact_codegen_runs(status);
        CREATE INDEX IF NOT EXISTS idx_fact_codegen_runs_created_at ON fact_codegen_runs(created_at);

        CREATE TABLE IF NOT EXISTS fact_codegen_generated_files (
          generated_file_id TEXT PRIMARY KEY NOT NULL,
          run_id TEXT NOT NULL,
          project_id TEXT,
          project_name TEXT,
          component_id TEXT,
          component_title TEXT,
          entity_name TEXT NOT NULL,
          file_path TEXT NOT NULL,
          language TEXT,
          file_size_bytes INTEGER,
          generation_order INTEGER NOT NULL,
          created_at TEXT NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_fact_codegen_generated_files_run_id ON fact_codegen_generated_files(run_id);
        CREATE INDEX IF NOT EXISTS idx_fact_codegen_generated_files_entity_name ON fact_codegen_generated_files(entity_name);

        CREATE TABLE IF NOT EXISTS fact_codegen_input_entities (
          input_entity_id TEXT PRIMARY KEY NOT NULL,
          run_id TEXT NOT NULL,
          project_id TEXT,
          project_name TEXT,
          component_id TEXT,
          component_title TEXT,
          source_causal_id TEXT,
          entity_name TEXT NOT NULL,
          source_head TEXT,
          source_relationship TEXT,
          source_tail TEXT,
          source_detail TEXT,
          created_at TEXT NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_fact_codegen_input_entities_run_id ON fact_codegen_input_entities(run_id);
        CREATE INDEX IF NOT EXISTS idx_fact_codegen_input_entities_entity_name ON fact_codegen_input_entities(entity_name);

        CREATE TABLE IF NOT EXISTS fact_codegen_metrics (
          metric_row_id TEXT PRIMARY KEY NOT NULL,
          run_id TEXT NOT NULL,
          project_id TEXT,
          project_name TEXT,
          component_id TEXT,
          component_title TEXT,
          metric_key TEXT NOT NULL,
          metric_type TEXT NOT NULL,
          metric_value TEXT NOT NULL,
          metric_number_value REAL,
          metric_boolean_value INTEGER,
          created_at TEXT NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_fact_codegen_metrics_run_id ON fact_codegen_metrics(run_id);
        CREATE INDEX IF NOT EXISTS idx_fact_codegen_metrics_key ON fact_codegen_metrics(metric_key);

        CREATE TABLE IF NOT EXISTS codegen_analytics_refresh_log (
          refresh_id INTEGER PRIMARY KEY AUTOINCREMENT,
          source_db_path TEXT NOT NULL,
          analytics_db_path TEXT NOT NULL,
          refreshed_at TEXT NOT NULL,
          fact_codegen_runs_count INTEGER NOT NULL,
          fact_codegen_generated_files_count INTEGER NOT NULL,
          fact_codegen_input_entities_count INTEGER NOT NULL,
          fact_codegen_metrics_count INTEGER NOT NULL,
          warnings_json TEXT NOT NULL
        );
        """
    )


def _drop_codegen_analytics_fact_tables(connection: sqlite3.Connection) -> None:
    connection.executescript(
        """
        DROP TABLE IF EXISTS fact_codegen_runs;
        DROP TABLE IF EXISTS fact_codegen_generated_files;
        DROP TABLE IF EXISTS fact_codegen_input_entities;
        DROP TABLE IF EXISTS fact_codegen_metrics;
        DROP TABLE IF EXISTS codegen_analytics_refresh_log;
        """
    )


def _clear_codegen_analytics_tables(connection: sqlite3.Connection) -> None:
    connection.executescript(
        """
        DELETE FROM fact_codegen_runs;
        DELETE FROM fact_codegen_generated_files;
        DELETE FROM fact_codegen_input_entities;
        DELETE FROM fact_codegen_metrics;
        """
    )


def _refresh_fact_codegen_runs(connection: sqlite3.Connection) -> None:
    connection.execute(
        """
        INSERT INTO fact_codegen_runs (
          run_id,
          project_id,
          project_name,
          component_id,
          component_title,
          component_category,
          causal_project_document_id,
          source_type,
          status,
          model,
          started_at,
          finished_at,
          duration_ms,
          duration_seconds,
          input_entity_count,
          generated_entity_count,
          generated_file_count,
          error_message,
          created_at,
          updated_at,
          is_success
        )
        SELECT
          run.id,
          run.project_id,
          project.name,
          run.component_id,
          component.title,
          component.category,
          run.causal_project_document_id,
          run.source_type,
          run.status,
          run.model,
          run.started_at,
          run.finished_at,
          run.duration_ms,
          CASE WHEN run.duration_ms IS NULL THEN NULL ELSE (run.duration_ms / 1000.0) END,
          run.input_entity_count,
          run.generated_entity_count,
          run.generated_file_count,
          run.error_message,
          run.created_at,
          run.updated_at,
          CASE WHEN run.status = 'completed' THEN 1 ELSE 0 END
        FROM src.codegen_runs AS run
        LEFT JOIN src.projects AS project ON project.id = run.project_id
        LEFT JOIN src.project_components AS component ON component.id = run.component_id;
        """
    )


def _refresh_fact_codegen_generated_files(connection: sqlite3.Connection) -> None:
    connection.execute(
        """
        INSERT INTO fact_codegen_generated_files (
          generated_file_id,
          run_id,
          project_id,
          project_name,
          component_id,
          component_title,
          entity_name,
          file_path,
          language,
          file_size_bytes,
          generation_order,
          created_at
        )
        SELECT
          file.id,
          run.id,
          run.project_id,
          project.name,
          run.component_id,
          component.title,
          file.entity_name,
          file.file_path,
          file.language,
          file.file_size_bytes,
          file.generation_order,
          file.created_at
        FROM src.codegen_generated_files AS file
        INNER JOIN src.codegen_runs AS run ON run.id = file.run_id
        LEFT JOIN src.projects AS project ON project.id = run.project_id
        LEFT JOIN src.project_components AS component ON component.id = run.component_id;
        """
    )


def _refresh_fact_codegen_input_entities(connection: sqlite3.Connection) -> None:
    connection.execute(
        """
        INSERT INTO fact_codegen_input_entities (
          input_entity_id,
          run_id,
          project_id,
          project_name,
          component_id,
          component_title,
          source_causal_id,
          entity_name,
          source_head,
          source_relationship,
          source_tail,
          source_detail,
          created_at
        )
        SELECT
          entity.id,
          run.id,
          run.project_id,
          project.name,
          run.component_id,
          component.title,
          entity.source_causal_id,
          entity.entity_name,
          entity.source_head,
          entity.source_relationship,
          entity.source_tail,
          entity.source_detail,
          entity.created_at
        FROM src.codegen_input_entities AS entity
        INNER JOIN src.codegen_runs AS run ON run.id = entity.run_id
        LEFT JOIN src.projects AS project ON project.id = run.project_id
        LEFT JOIN src.project_components AS component ON component.id = run.component_id;
        """
    )


def _refresh_fact_codegen_metrics(connection: sqlite3.Connection) -> None:
    connection.execute(
        """
        INSERT INTO fact_codegen_metrics (
          metric_row_id,
          run_id,
          project_id,
          project_name,
          component_id,
          component_title,
          metric_key,
          metric_type,
          metric_value,
          metric_number_value,
          metric_boolean_value,
          created_at
        )
        SELECT
          metric.id,
          run.id,
          run.project_id,
          project.name,
          run.component_id,
          component.title,
          metric.metric_key,
          metric.metric_type,
          metric.metric_value,
          CASE WHEN metric.metric_type = 'number' THEN CAST(metric.metric_value AS REAL) ELSE NULL END,
          CASE
            WHEN metric.metric_type = 'boolean' AND lower(metric.metric_value) IN ('1', 'true', 'yes') THEN 1
            WHEN metric.metric_type = 'boolean' AND lower(metric.metric_value) IN ('0', 'false', 'no') THEN 0
            ELSE NULL
          END,
          metric.created_at
        FROM src.codegen_run_metrics AS metric
        INNER JOIN src.codegen_runs AS run ON run.id = metric.run_id
        LEFT JOIN src.projects AS project ON project.id = run.project_id
        LEFT JOIN src.project_components AS component ON component.id = run.component_id;
        """
    )


def _table_row_count(connection: sqlite3.Connection, table_name: str) -> int:
    row = connection.execute(f"SELECT COUNT(*) FROM {table_name}").fetchone()
    return int(row[0]) if row else 0


def refresh_codegen_analytics(
    source_db_path: str | None = None,
    analytics_db_path: str | None = None,
) -> dict[str, Any]:
    source_path = _resolve_source_db_path(source_db_path)
    analytics_path = _resolve_analytics_db_path(analytics_db_path)

    if not source_path.exists():
        raise CodegenAnalyticsError(
            f"Source database does not exist: {source_path}. "
            "Run web-ui first so local.db is created."
        )

    analytics_path.parent.mkdir(parents=True, exist_ok=True)

    warnings: list[str] = []

    with sqlite3.connect(source_path) as source_connection, sqlite3.connect(analytics_path) as analytics_connection:
        analytics_connection.execute("PRAGMA foreign_keys = ON")
        analytics_connection.execute("ATTACH DATABASE ? AS src", (str(source_path),))

        try:
            _drop_codegen_analytics_fact_tables(analytics_connection)
            _create_codegen_analytics_schema(analytics_connection)
            _clear_codegen_analytics_tables(analytics_connection)

            missing_source_tables = [
                table_name
                for table_name in CODEGEN_SOURCE_TABLES
                if not _table_exists(analytics_connection, table_name, schema_name="src")
            ]

            if missing_source_tables:
                warnings.append(
                    "Missing source tables: "
                    + ", ".join(sorted(missing_source_tables))
                    + ". Run latest drizzle migrations in web-ui to enable codegen analytics ingestion."
                )
            else:
                _refresh_fact_codegen_runs(analytics_connection)
                _refresh_fact_codegen_generated_files(analytics_connection)
                _refresh_fact_codegen_input_entities(analytics_connection)
                _refresh_fact_codegen_metrics(analytics_connection)

            counts = {table_name: _table_row_count(analytics_connection, table_name) for table_name in CODEGEN_ANALYTICS_TABLES}
            refreshed_at = datetime.now(timezone.utc).isoformat()

            analytics_connection.execute(
                """
                INSERT INTO codegen_analytics_refresh_log (
                  source_db_path,
                  analytics_db_path,
                  refreshed_at,
                  fact_codegen_runs_count,
                  fact_codegen_generated_files_count,
                  fact_codegen_input_entities_count,
                  fact_codegen_metrics_count,
                  warnings_json
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    str(source_path),
                    str(analytics_path),
                    refreshed_at,
                    counts["fact_codegen_runs"],
                    counts["fact_codegen_generated_files"],
                    counts["fact_codegen_input_entities"],
                    counts["fact_codegen_metrics"],
                    json.dumps(warnings),
                ),
            )

            analytics_connection.commit()

            return {
                "sourceDbPath": str(source_path),
                "analyticsDbPath": str(analytics_path),
                "refreshedAt": refreshed_at,
                "counts": counts,
                "warnings": warnings,
            }
        finally:
            try:
                analytics_connection.execute("DETACH DATABASE src")
            except sqlite3.Error:
                pass


def list_codegen_analytics_tables(analytics_db_path: str | None = None) -> list[dict[str, Any]]:
    analytics_path = _resolve_analytics_db_path(analytics_db_path)
    if not analytics_path.exists():
        raise CodegenAnalyticsError(
            f"Analytics database does not exist: {analytics_path}. "
            "Run POST /analytics/codegen/refresh first."
        )

    with sqlite3.connect(analytics_path) as connection:
        return [
            {"name": table_name, "rows": _table_row_count(connection, table_name)}
            for table_name in CODEGEN_ANALYTICS_TABLES
        ]


def query_codegen_analytics_table(
    table_name: str,
    limit: int = 1000,
    offset: int = 0,
    analytics_db_path: str | None = None,
) -> dict[str, Any]:
    if table_name not in CODEGEN_ANALYTICS_TABLES:
        allowed = ", ".join(CODEGEN_ANALYTICS_TABLES)
        raise CodegenAnalyticsError(f"Unsupported table '{table_name}'. Allowed values: {allowed}")

    safe_limit = max(1, min(limit, 10000))
    safe_offset = max(0, offset)

    analytics_path = _resolve_analytics_db_path(analytics_db_path)
    if not analytics_path.exists():
        raise CodegenAnalyticsError(
            f"Analytics database does not exist: {analytics_path}. "
            "Run POST /analytics/codegen/refresh first."
        )

    with sqlite3.connect(analytics_path) as connection:
        connection.row_factory = sqlite3.Row
        total_row = connection.execute(f"SELECT COUNT(*) AS count FROM {table_name}").fetchone()
        total = int(total_row["count"]) if total_row else 0

        rows = connection.execute(
            f"SELECT * FROM {table_name} LIMIT ? OFFSET ?",
            (safe_limit, safe_offset),
        ).fetchall()

    return {
        "table": table_name,
        "limit": safe_limit,
        "offset": safe_offset,
        "total": total,
        "rows": [dict(row) for row in rows],
    }
